subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStep.java

```diff
@@ -0,0 +1,227 @@
+        work.visitInputProperties((propertyName, value) -> {
+            try {
+                ValueSnapshot previousSnapshot = previousSnapshots.get(propertyName);
+                if (previousSnapshot == null) {
+                    builder.put(propertyName, valueSnapshotter.snapshot(value));
+                } else {
+                    builder.put(propertyName, valueSnapshotter.snapshot(value, previousSnapshot));
+                }
+            } catch (Exception e) {
+                throw new UncheckedIOException(String.format("Unable to store input properties for %s. Property '%s' with value '%s' cannot be serialized.",
+                    work.getDisplayName(), propertyName, value), e);
+            }
+        });
+
+        return builder.build();
+    }
+
+    private static ImmutableSortedMap<String, CurrentFileCollectionFingerprint> fingerprintInputFiles(UnitOfWork work) {
+        ImmutableSortedMap.Builder<String, CurrentFileCollectionFingerprint> builder = ImmutableSortedMap.naturalOrder();
+        work.visitInputFileProperties((propertyName, value, incremental, fingerprinter) -> {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Fingerprinting property {} for {}", propertyName, work.getDisplayName());
+            }
+            CurrentFileCollectionFingerprint result = fingerprinter.get();
+            builder.put(propertyName, result);
+        });
+        return builder.build();
+    }
+
+    private static ImmutableSortedMap<String, CurrentFileCollectionFingerprint> fingerprintOutputFiles(
+        ImmutableSortedMap<String, FileCollectionFingerprint> previousOutputFingerprints,
+        ImmutableSortedMap<String, FileSystemSnapshot> beforeExecutionOutputSnapshots,
+        boolean hasOverlappingOutputs
+    ) {
+        return ImmutableSortedMap.copyOfSorted(
+            Maps.transformEntries(beforeExecutionOutputSnapshots, (key, outputSnapshot) -> {
+                    FileCollectionFingerprint previousOutputFingerprint = previousOutputFingerprints.get(key);
+                    //noinspection ConstantConditions
+                    return previousOutputFingerprint == null
+                        ? AbsolutePathFingerprintingStrategy.IGNORE_MISSING.getEmptyFingerprint()
+                        : fingerprintOutputSnapshot(outputSnapshot, previousOutputFingerprint, hasOverlappingOutputs);
+                }
+            )
+        );
+    }
+
+    private static CurrentFileCollectionFingerprint fingerprintOutputSnapshot(FileSystemSnapshot beforeExecutionOutputSnapshots, FileCollectionFingerprint previousOutputFingerprint, boolean hasOverlappingOutputs) {
+        List<FileSystemSnapshot> roots = hasOverlappingOutputs
+            ? OutputFilterUtil.filterOutputSnapshotBeforeExecution(previousOutputFingerprint, beforeExecutionOutputSnapshots)
+            : ImmutableList.of(beforeExecutionOutputSnapshots);
```

Comment:
We probably should use a `CompositeFileSystemSnapshot` here as well instead of passing around a list.

Reply:
Yes, though I'd like to address this in the next PR where we extract this (currently duplicated) code to another shared execution step.